File: sdk/python/examples/follow_feed.py
"""Page through a feed generator until exhausted or until a page cap is hit.

Run:
    uv run python examples/follow_feed.py at://did:plc:.../app.rocksky.feed.generator/all
"""

from __future__ import annotations

import asyncio
import sys

from rocksky import Client


async def main(feed_uri: str, max_pages: int = 3) -> int:
    cursor: str | None = None
    seen = 0
    async with Client() as rocksky:
        for page in range(max_pages):
            feed = await rocksky.feed.get(feed_uri, limit=25, cursor=cursor)
            for item in feed.feed:
                s = item.scrobble
                if not s:
                    continue
                seen += 1
                stamp = s.date.isoformat() if s.date else "—"
                print(f"  [{stamp}]  {s.user:24s}  {s.artist} — {s.title}")
            cursor = feed.cursor
            print(f"--- end of page {page + 1}; cursor={cursor!r} ---")
            if not cursor:
                break
    print(f"\ntotal items: {seen}")
    return 0


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("usage: follow_feed.py <feed-at-uri>")
        sys.exit(2)
    sys.exit(asyncio.run(main(sys.argv[1])))

File: sdk/rust/examples/follow_feed.rs
//! Page through the authenticated user's "following" scrobble feed.
//!
//! ```text
//! ROCKSKY_TOKEN=... cargo run --example follow_feed -- 50
//! ```

use rocksky::Client;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let total = std::env::args()
        .nth(1)
        .and_then(|s| s.parse::<u32>().ok())
        .unwrap_or(20);

    let Ok(token) = std::env::var("ROCKSKY_TOKEN") else {
        eprintln!("set $ROCKSKY_TOKEN first");
        std::process::exit(1);
    };
    let client = Client::builder().token(token).build();

    let page = 20;
    let mut fetched = 0u32;
    let mut offset = 0u32;

    while fetched < total {
        let want = (total - fetched).min(page);
        let scrobbles = client
            .scrobble()
            .list()
            .following(true)
            .limit(want)
            .offset(offset)
            .send()
            .await?;
        if scrobbles.is_empty() {
            break;
        }
        for s in &scrobbles {
            let when = s
                .date
                .map(|d| d.to_rfc3339())
                .unwrap_or_else(|| "—".into());
            println!(
                "[{when}] {} — {} (by {})",
                s.artist.clone().unwrap_or_else(|| "?".into()),
                s.title.clone().unwrap_or_else(|| "?".into()),
                s.user.clone().unwrap_or_else(|| "?".into()),
            );
        }
        let got = scrobbles.len() as u32;
        fetched += got;
        offset += got;
        if got < want {
            break;
        }
    }
    eprintln!("\nfetched {fetched} scrobble(s)");
    Ok(())
}
